e2e/testsuite/testsuite.go

```diff
@@ -0,0 +1,286 @@
+package testsuite
+
+import (
+	"context"
+	"fmt"
+	"io/ioutil"
+	"strings"
+	"testing"
+	"time"
+
+	dockerclient "github.com/docker/docker/client"
+	"github.com/strangelove-ventures/ibctest"
+	"github.com/strangelove-ventures/ibctest/chain/cosmos"
+	"github.com/strangelove-ventures/ibctest/ibc"
+	"github.com/strangelove-ventures/ibctest/relayer"
+	"github.com/strangelove-ventures/ibctest/testreporter"
+	"github.com/stretchr/testify/suite"
+	"go.uber.org/zap"
+	"go.uber.org/zap/zaptest"
+
+	"github.com/cosmos/ibc-go/v4/e2e/testconfig"
+)
+
+const (
+	SourceRelayerName      = "srcRly"
+	DestinationRelayerName = "dstRly"
+)
+
+// E2ETestSuite has methods and functionality which can be shared among all test suites.
+type E2ETestSuite struct {
+	suite.Suite
+	chainPairs     map[string]chainPair
+	logger         *zap.Logger
+	Client         *dockerclient.Client
+	network        string
+	startRelayerFn func(relayer ibc.Relayer)
+}
+
+// chainPair is a pairing of two chains which will be used in a test.
+type chainPair struct {
+	srcChain, dstChain *cosmos.CosmosChain
+}
+
+// ChainOptions stores chain configurations for the chains that will be
+// created for the tests. They can be modified by passing ChainOptionConfiguration
+// to E2ETestSuite.GetChains.
+type ChainOptions struct {
+	SrcChainConfig *ibc.ChainConfig
+	DstChainConfig *ibc.ChainConfig
+}
+
+// ChainOptionConfiguration enables arbitrary configuration of ChainOptions.
+type ChainOptionConfiguration func(options *ChainOptions)
+
+// GetChains returns a src and dst chain that can be used in a test. The pair returned
+// is unique to the current test being run. Note: this function does not create containers.
+func (s *E2ETestSuite) GetChains(chainOpts ...ChainOptionConfiguration) (*cosmos.CosmosChain, *cosmos.CosmosChain) {
+	if s.chainPairs == nil {
+		s.chainPairs = map[string]chainPair{}
+	}
+	cp, ok := s.chainPairs[s.T().Name()]
+	if ok {
+		return cp.srcChain, cp.dstChain
+	}
+
+	chainOptions := defaultChainOptions()
+	for _, opt := range chainOpts {
+		opt(&chainOptions)
+	}
+
+	srcChain, dstChain := s.createCosmosChains(chainOptions)
+	cp = chainPair{
+		srcChain: srcChain,
+		dstChain: dstChain,
+	}
+	s.chainPairs[s.T().Name()] = cp
+
+	return cp.srcChain, cp.dstChain
+}
+
+// CreateChainsRelayerAndChannel create two chains, a relayer, establishes a connection and creates a channel
+// using the given channel options. The relayer returned by this function has not yet started. It should be started
+// with E2ETestSuite.StartRelayer if needed.
+// This should be called at the start of every test, unless fine grained control is required.
+func (s *E2ETestSuite) CreateChainsRelayerAndChannel(ctx context.Context, channelOpts ...func(*ibc.CreateChannelOptions)) ibc.Relayer {
```

Comment:
I'd rename to `Setup`. I also suggest making this the first function after type declarations